combinetf2/fitter.py

```diff
@@ -0,0 +1,1410 @@
+        nstat = self.npoi + self.indata.nsystnoconstraint
+        hess_stat = self.hess[:nstat, :nstat]
+        identity = tf.eye(nstat, dtype=hess_stat.dtype)
+        inv_hess_stat = tf.linalg.solve(hess_stat, identity)  # Solves H * X = I
+
+        if self.binByBinStat:
+            # impact bin-by-bin stat
+            val_no_bbb, grad_no_bbb, hess_no_bbb = self.loss_val_grad_hess(
+                profile_grad=False
+            )
+
+            hess_stat_no_bbb = hess_no_bbb[:nstat, :nstat]
+            inv_hess_stat_no_bbb = tf.linalg.solve(hess_stat_no_bbb, identity)
+
+            impacts_data_stat = tf.sqrt(tf.linalg.diag_part(inv_hess_stat_no_bbb))
+            impacts_data_stat = tf.reshape(impacts_data_stat, (-1, 1))
+
+            impacts_bbb_sq = tf.linalg.diag_part(inv_hess_stat - inv_hess_stat_no_bbb)
+            impacts_bbb = tf.sqrt(tf.nn.relu(impacts_bbb_sq))  # max(0,x)
+            impacts_bbb = tf.reshape(impacts_bbb, (-1, 1))
+            impacts_grouped = tf.concat([impacts_data_stat, impacts_bbb], axis=1)
+        else:
+            impacts_data_stat = tf.sqrt(tf.linalg.diag_part(inv_hess_stat))
+            impacts_data_stat = tf.reshape(impacts_data_stat, (-1, 1))
+            impacts_grouped = impacts_data_stat
+
+        if len(self.indata.systgroupidxs):
+            impacts_grouped_syst = tf.map_fn(
+                lambda idxs: self._compute_impact_group(v[:, self.npoi :], idxs),
+                tf.ragged.constant(self.indata.systgroupidxs, dtype=tf.int32),
+                fn_output_signature=tf.TensorSpec(
+                    shape=(impacts.shape[0],), dtype=tf.float64
+                ),
+            )
+            impacts_grouped_syst = tf.transpose(impacts_grouped_syst)
+            impacts_grouped = tf.concat([impacts_grouped_syst, impacts_grouped], axis=1)
+
+        return impacts, impacts_grouped
+
+    def _compute_global_impact_group(self, d_squared, idxs):
+        gathered = tf.gather(d_squared, idxs, axis=-1)
+        d_squared_summed = tf.reduce_sum(gathered, axis=-1)
+        return tf.sqrt(d_squared_summed)
+
+    @tf.function
+    def global_impacts_parms(self):
+        # compute impacts for pois and nois
+        dxdtheta0, dxdnobs, dxdbeta0 = self._compute_derivatives_x()
+
+        dxdtheta0_poi = dxdtheta0[: self.npoi]
+        dxdtheta0_noi = tf.gather(dxdtheta0[self.npoi :], self.indata.noigroupidxs)
+        dxdtheta0 = tf.concat([dxdtheta0_poi, dxdtheta0_noi], axis=0)
+        dxdtheta0_squared = tf.square(dxdtheta0)
+
+        # global impact data stat
+        dxdnobs_poi = dxdnobs[: self.npoi]
+        dxdnobs_noi = tf.gather(dxdnobs[self.npoi :], self.indata.noigroupidxs)
+        dxdnobs = tf.concat([dxdnobs_poi, dxdnobs_noi], axis=0)
+
+        if self.externalCovariance:
+            data_cov = tf.linalg.inv(self.data_cov_inv)
+            data_stat = tf.einsum("ij,jk,ik->i", dxdnobs, data_cov, dxdnobs)
```

Comment:
Can you at least add a comment with the corresponding matrix multiplication formula?
(it's dxdnobs @ data_cov @ dxdnobs.T I guess?)
Is the einsum really better/faster?

Reply:
The equivalent would be ```tf.linalg.diag_part(dxdnobs @ data_cov @ tf.transpose(dxdnobs))```. The einsum should be faster since it avoids computing the full matrix, this is also what I see when I do a quick benchmarking.